### PR TITLE
GH-100112:  avoid using iterable coroutines in asyncio internally

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-100112

### CODE REVIEW

**Review by gvanrossum — COMMENTED** [review]
So as you've seen from my comment on the issue, I don't understand what's going on here. Why do we need to add this? Since `_wrap_awaitable` is internal anyway, and there's only one caller, maybe we can simplify this by modifying the call site? (FWIW, the variable `called_wrap_awaitable` at that call site should probably be called `should_close` with the truty value flipped, to make it clearer what its purpose is.

**Comment by kumaraditya303** [comment]
> So as you've seen from my comment on the issue, I don't understand what's going on here. Why do we need to add this? Since _wrap_awaitable is internal anyway, and there's only one caller, maybe we can simplify this by modifying the call site? (FWIW, the variable called_wrap_awaitable at that call site should probably be called should_close with the truty value flipped, to make it clearer what its purpose is.

Yeah, it will be better to inline it. I'll write some tests for this first before proceeding here. (this should have been a draft)

File: Lib/asyncio/tasks.py [hunk 1]
```diff
@@ -635,11 +634,17 @@ def ensure_future(coro_or_future, *, loop=None):
...
         return coro_or_future
-    called_wrap_awaitable = False
+    should_close = False
     if not coroutines.iscoroutine(coro_or_future):
         if inspect.isawaitable(coro_or_future):
+            async def _wrap_awaitable(awaitable):
+                @types.coroutine
+                def wrapper():
+                    return (yield from awaitable.__await__())
+                return await wrapper()
```
Comment (graingert) [inline Lib/asyncio/tasks.py]:
```suggestion
                return await awaitable
```

or is the aim to support:

`obj.__await__ = lambda: asyncio.sleep(0, result='ok').__await__()`?

**Review by gvanrossum — APPROVED** [review]
Thanks -- I finally understand the issue (thanks for the clarification and the example) and this looks right.

**Pull request merged**